File: server/src/websocket.ts
import { Frame, FrameHandler, WebSocketAddress, Address, LinkBroadcastError, LinkSendError, Protocol } from "@core/net";
import * as WebSocketFrame from "@core/media/websocket";
import { WebSocketServer, WebSocket } from "ws";
import { Result } from "oxide.ts";

class WebSocketConnection implements WebSocketFrame.Connection {
    #socket: WebSocket;
    #remote: WebSocketAddress;

    constructor(args: { socket: WebSocket; remote: WebSocketAddress }) {
        this.#socket = args.socket;
        this.#remote = args.remote;
    }

    get remote(): WebSocketAddress {
        return this.#remote;
    }

    send(buffer: Uint8Array): void {
        this.#socket.send(buffer);
    }

    onReceive(callback: (buffer: Uint8Array) => void): void {
        this.#socket.on("message", (data) => {
            if (!(data instanceof Buffer)) {
                console.warn("WebSocket received non-buffer data", data);
                return;
            }
            callback(data);
        });
    }

    onClose(callback: () => void): void {
        this.#socket.on("close", callback);
    }
}

export class WebSocketHandler implements FrameHandler {
    #inner: WebSocketFrame.WebSocketHandler;

    constructor(args: { port: number }) {
        this.#inner = new WebSocketFrame.WebSocketHandler();
        const server = new WebSocketServer({ port: args.port });
        server.on("connection", (socket, req) => {
            const { remoteAddress, remotePort } = req.socket;
            if (remoteAddress === undefined || remotePort === undefined) {
                return;
            }
            console.log(`WebSocket connection from ${remoteAddress}:${remotePort}`);
            const remote = WebSocketAddress.schema.parse(`${remoteAddress}:${remotePort}`);
            this.#inner.addConnection(new WebSocketConnection({ socket, remote }));
        });
        server.on("listening", () => {
            console.log(`WebSocket server listening on port ${args.port}`);
        });
    }

    address(): Address | undefined {
        return this.#inner.address();
    }

    send(frame: Frame): Result<void, LinkSendError> {
        return this.#inner.send(frame);
    }

    sendBroadcast?(protocol: Protocol, payload: Uint8Array): Result<void, LinkBroadcastError> {
        return this.#inner.sendBroadcast(protocol, payload);
    }

    onReceive(callback: (frame: Frame) => void): void {
        this.#inner.onReceive(callback);
    }

    onClose(callback: () => void): void {
        this.#inner.onClose(callback);
    }
}

File: py-core/py_core/media/websocket.py
import asyncio
from dataclasses import dataclass
from ipaddress import IPv4Address
import logging
from typing import Final, Self
import websockets
from websockets.client import WebSocketClientProtocol
from websockets.server import WebSocketServerProtocol, serve

from py_core.net.link.address import WebSocketAddress
from py_core.net.link.frame import LinkFrame, Protocol, ReceivedLinkFrame
from py_core.net.link.service import FrameHandler
from py_core.serde.buffer import BufferReader, BufferWriter
from py_core.serde.bytes import RemainingBytes
from py_core.serde.derive import DeriveDeserialize, DeriveSerialize

WEBSOCKET_PORT: Final[int] = 12346


@dataclass
class WebSocketFrame(DeriveDeserialize, DeriveSerialize):
    protocol: Protocol
    payload: RemainingBytes

    @classmethod
    def from_link_frame(cls, frame: LinkFrame) -> Self:
        return cls(frame.protocol, RemainingBytes(frame.payload))

    def into_received_link_frame(
        self, address: WebSocketAddress, disconnect_event: asyncio.Event
    ) -> ReceivedLinkFrame:
        return ReceivedLinkFrame(
            self.protocol, address, self.payload.value, disconnect_event
        )


@dataclass
class RawPacket:
    data: bytes
    address: WebSocketAddress
    disconnect_event: asyncio.Event


type WebSocketProtocol = WebSocketServerProtocol | WebSocketClientProtocol


class WebSocketHandler(FrameHandler):
    _server: serve
    _connections: dict[WebSocketAddress, WebSocketProtocol]
    _recv_queue: asyncio.Queue[RawPacket]

    async def _process_protocol(self, ws: WebSocketProtocol):
        host, port = ws.remote_address
        addresss = WebSocketAddress(IPv4Address(host), port)
        self._connections[addresss] = ws

        disconnect_event = asyncio.Event()

        async def remove_connection():
            await ws.wait_closed()
            del self._connections[addresss]
            disconnect_event.set()

        asyncio.create_task(remove_connection())

        async def recv():
            async for data in ws:
                if isinstance(data, bytes):
                    packet = RawPacket(data, addresss, disconnect_event)
                    await self._recv_queue.put(packet)
                else:
                    logging.warning(f"Received non-bytes data: {data}")

        asyncio.create_task(recv())

    def __init__(self):
        self._server = serve(self._process_protocol, "0.0.0.0", WEBSOCKET_PORT)

    async def connect(self, address: WebSocketAddress, use_ssl=False) -> None:
        scheme = "wss" if use_ssl else "ws"
        uri = f"{scheme}://{address.host}:{address.port}"
        async with websockets.connect(uri) as ws:
            asyncio.create_task(self._process_protocol(ws))

    async def recv(self) -> ReceivedLinkFrame:
        packet = await self._recv_queue.get()
        reader = BufferReader(packet.data)
        frame = WebSocketFrame.deserialize(reader)
        return frame.into_received_link_frame(packet.address, packet.disconnect_event)

    async def send(self, frame: LinkFrame) -> None:
        if not isinstance(frame.address, WebSocketAddress):
            raise ValueError(f"Invalid address: {frame.address}")

        connection = self._connections.get(frame.address)
        if connection is not None:
            ws_frame = WebSocketFrame.from_link_frame(frame)
            writer = BufferWriter(ws_frame.serialized_length())
            await connection.send(writer.unwrap_buffer())

    async def send_broadcast(self, protocol: Protocol, payload: bytes) -> None:
        for address in self._connections.keys():
            frame = LinkFrame(protocol, address, payload)
            await self.send(frame)

    async def close(self) -> None:
        self._server.ws_server.close()
        await self._server.ws_server.wait_closed()
